File: android/app/src/main/kotlin/org/hyn/titan/sensor/SensorManager.kt
package org.hyn.titan.sensor

import android.content.Context
import android.os.Build
import timber.log.Timber

class SensorManager(val context: Context, private val sensorValueChangeListener: OnSensorValueChangeListener) {


    var registerSensorList = mutableListOf<Sensor>()


//    var sensorValueChangeListener: OnSensorValueChangeListener = object : OnSensorValueChangeListener {
//        override fun onSensorChange(sensorType: Int, values: Map<String, Any>) {
//            Timber.i("sensorType:$sensorType,values:$values")
//        }
//    }

    fun init() {
        registerSensorList.clear()
        val wifiSensor = WifiSensor(context, sensorValueChangeListener);
        registerSensorList.add(wifiSensor)
        val blueToothSensor = BluetoothSensor(context, sensorValueChangeListener);
        registerSensorList.add(blueToothSensor)
        val gpsSensor = GpsSensor(context, sensorValueChangeListener);
        registerSensorList.add(gpsSensor)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            val gnssSensor = GnssSensor(context, sensorValueChangeListener);
            registerSensorList.add(gnssSensor)
        }

        val cellularSensor = CellularSensor(context, sensorValueChangeListener);

        registerSensorList.add(cellularSensor)

    }


    fun startScan() {
        for (sensor in registerSensorList) {
            sensor.startScan();
        }
    }

    fun stopScan() {
        for (sensor in registerSensorList) {
            sensor.stopScan();
        }
    }

    fun destory() {
        for (sensor in registerSensorList) {
            sensor.destory();
        }
        registerSensorList.clear()
    }
}

File: ios/Runner/Sensor/SensorManager.swift
//
//  SensorManager.swift
//  Runner
//
//  Created by naru.j on 2019/12/20.
//  Copyright © 2019 The Chromium Authors. All rights reserved.
//

import Foundation

class SensorManager {
    
    var registerSensorList = Array<Sensor>()
    var onSensorChange: OnSensorValueChangeListener!

    func initialize() {
        registerSensorList.removeAll()

        let blueToothSensor = BluetoothSensor();
        blueToothSensor.onSensorChange = onSensorChange
        registerSensorList.append(blueToothSensor)
        
        let gpsSensor = GpsSensor()
        gpsSensor.onSensorChange = onSensorChange
//        registerSensorList.append(gpsSensor)
        
        let cellularSensor = CellularSensor()
        cellularSensor.onSensorChange = onSensorChange
//        registerSensorList.append(cellularSensor)

        for sensor in registerSensorList {
            sensor.initialize()
        }
    }
    
    func startScan() {
        for sensor in registerSensorList {
            sensor.startScan()
        }
        print("[ios] --> startScan")
    }

    func stopScan() {
        for sensor in registerSensorList {
            sensor.stopScan()
        }
        print("[ios] --> stopScan")
    }

    func destory() {
        for sensor in registerSensorList {
            sensor.destory();
        }
        registerSensorList.removeAll()
        
        print("[ios] --> destory")
    }
}
